### PR TITLE
Vertx3 route handling

### PR DESCRIPTION
**Pull request by zingmane**
Switched from vertx2 router handling to vertx3 so it's easier to handle cookies and auth

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -19,8 +19,8 @@ plugins {
 project.group = 'com.campudus'
 project.version = '0.1.0'
 
-def vertxVersion = '3.5.4'  // 3.6.0 is out
-def scalaVersion = '2.12.7' // 2.12.8 is out
+def vertxVersion = '3.6.3'
+def scalaVersion = '2.12.6'
```
Comment (Exotica) [inline build.gradle]: Aus Interesse: Warum nicht 2.12.8?

Comment (zingmane) [reply build.gradle]: Hm, no particular background, probably a typo.

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 2]
```diff
@@ -69,35 +77,29 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
-                p.success(r)
+        case _ =>
+          Try(Json.fromObjectString(buffer)) match {
+            case Success(r) => r
 
-              case Failure(ex: DecodeException) =>
-                logger.error(s"Couldn't parse requestBody. JSON is valid: [$requestBody]")
-                p.failure(InvalidJsonException(ex.getMessage, "invalid"))
+            case Failure(ex: DecodeException) =>
+              logger.error(s"Couldn't parse requestBody. JSON is valid: [$buffer]")
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: 👍

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 3]
```diff
@@ -69,35 +77,29 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+              throw InvalidJsonException(ex.getMessage, "invalid")
 
-              case Failure(ex) =>
-                logger.error(s"Couldn't parse requestBody. Excepted JSON but got: [$requestBody]")
-                p.failure(ex)
-            }
-        }
+            case Failure(ex) =>
+              logger.error(s"Couldn't parse requestBody. Excepted JSON but got: [$buffer]")
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: 👍

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 4]
```diff
@@ -116,4 +118,204 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+      .fileSystem()
+      .existsFuture(file)
+      .flatMap({
+        case true => Future.successful(file)
+        case false => Future.failed(new FileNotFoundException(file))
+      })
+  }
+
+  private def addIndexToDirName(path: String): String = {
+    if (path.endsWith("/")) path + "index.html"
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: No, they don't. That wasn't styled after moving it from old Router class.

[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: No, they don't. That wasn't styled after moving it from old Router class.

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 5]
```diff
@@ -116,4 +118,204 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+              true
+            case "gif" =>
+              resp.putHeader("Content-type", "image/gif")
+              true
+            case _ | "txt" =>
+              resp.putHeader("Content-type", "text/plain; charset= UTF-8")
+              false
+          }
+
+          val is = getClass.getResourceAsStream(path)
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: It stands for ìnputStream`. This is how the variable should be named.

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 6]
```diff
@@ -116,4 +118,204 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+          }
+        } catch {
+          case NonFatal(ex) =>
+            endResponse(
+              resp,
+              Error(RouterException("send embedded file exception", ex, "errors.routing.sendEmbeddedFile", 500)))
+        }
+      case SendFile(path, absolute) =>
+        (for {
+          exists <- fileExists(if (absolute) path else s"$workingDirectory/$path")
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: You're right. This is neither robust nor platfor agnostic. I change it to path concatenation with `java.io.File`

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 7]
```diff
@@ -116,4 +118,204 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+          file <- directoryToIndexFile(exists)
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: Agree. This was also moved from the old `Router`. I'll refactore it.

File: src/main/scala/com/campudus/tableaux/router/BaseRouter.scala [hunk 8]
```diff
@@ -116,4 +118,204 @@ trait BaseRouter extends Router with VertxAccess with LazyLogging {
...
+          resp.setStatusCode(200)
+          resp.setStatusMessage("OK")
+
+          val extension = if (path.contains(".")) {
+            path.split("\\.").toList.lastOption.map(_.toLowerCase)
+          } else {
+            "other"
+          }
+
+          val byteResponse = extension match {
```
[the remark this answers is not in the document]
Comment (zingmane) [reply src/main/scala/com/campudus/tableaux/router/BaseRouter.scala]: Totally agree

**Pull request merged**